Automattic/jetpack
# Fix Stripe nudge autosave and redirects in FSE

* The existing redirects after Stripe connection only work in the block editor. In the site editor, the redirect always falls back to the `earn/payments` settings page. The templates also fail to auto-save before you are taken to the Stripe connection form.
* The Recurring Payments block in particular was also not redirecting correctly in the _block-editor_, and thus also fell back to the settings page.

#### Changes proposed in this Pull Request:

* Refactors the logic which modifies the Stripe connection URL to redirect correctly out into a shared function
    * Adds state to the redirect url, which will be used by the API to redirect back to the site editor. This implementation chosen to avoid disrupting how the redirects currently work for the block editor.
* Refactors autosave and redirection logic, introduced to upgrade nudges in #19636 to handle both site and block editor, out into a reusable hook
* Updates the shared Stripe nudge to use both of these new shared functions (this is the old-style nudge seen in the Donations block, see screenshots)
* Updates the `Connect Stripe` button in the toolbars for Premium Content and Recurring Payments to use these new shared functions


**Not included: API change to actually redirect correctly in the site editor. SEE D60831**

#### Jetpack product discussion



#### Does this pull request change what data or activity we track or use?



#### Testing instructions:






You will need a test site with a blocks theme, premium plan and without a Stripe connection.
* Between each test of the Stripe nudge, you will need to disconnect the Stripe account you've just added from the `earn/payments` settings page. You may need to refresh this page a few times for the option to disconnect to appear.
* Because this PR does not include the backend changes to redirect to the site-editor, you should only test the autosave feature, and that it does not break existing redirects. The Upgrade nudges need to be regression tested due to the refactor.
* Tests are done with Premium Content, Recurring Payments, and Donations because the refactor touches these three blocks differently.

**Test Stripe Nudges**

1. Within the site editor (`<blog_url>/wp-admin/admin.php?page=gutenberg-edit-site`), insert a Payments block. Make a few modifications to the page, including to other template parts. Note the changes you've made, but do not save.
2. Press the `Connect Stripe` button in the block toolbar. Note that the Save icon in the top right of the page shows that it is in a saving state.
3. Complete or skip (if sandboxed) the Stripe connection form. Verify you are redirected to the **`earn/payments` page for now.**
4. Verify that the changes you made prior to connecting were saved.
5. Repeat steps 1-4 with a Premium Content and Donations block. (For Donations, the button will be in a banner, rather than the block toolbar).
6. Repeat steps 1- 4 using at least one of the blocks within the Calypso iframe (ie visit the site-editor at `wordpress.com/site-editor/<blog_url>`)
7. Within the **block editor**, insert a Payments block. Make a few modifications to the post before using the block to Connect to Stripe. Verify you are redirected to the post in the block editor, and your changes were saved
8. Repeat test 6 with a Premium Content and Donations block.

**Test Upgrade Nudges**

Repeat all tests from #19636
